l5kit/l5kit/environment/envs/l5_env.py

```diff
@@ -0,0 +1,185 @@
+
+        # Define action and observation space
+        # Continuous Action Space: gym.spaces.Box (X, Y, Yaw * number of future states)
+        # self.action_space = spaces.Box(low=-1000, high=1000, shape=(self.future_num_frames * 3, ))
+        self.action_space = spaces.Box(low=-1, high=1, shape=(self.future_num_frames * 3, ))
+
+        # Observation Space: gym.spaces.Dict (image: [n_channels, raster_size, raster_size])
+        self.observation_space = spaces.Dict({'image': spaces.Box(low=0, high=1,
+                                              shape=(n_channels, raster_size, raster_size), dtype=np.float32)})
+
+        # Simulator Config within Gym
+        self.sim_cfg = sim_cfg
+        self.simulator = ClosedLoopSimulator(self.sim_cfg, self.dataset, device=torch.device("cpu"),
+                                             verify_model=False)
+
+        # Reward
+        self.reward = reward
+
+        # Scene Sampling config
+        self.max_scene_id = cfg["gym_params"]["max_scene_id"]
+        if cfg["gym_params"]["overfit"]:  # Overfit
+            self.max_scene_id = 0
+
+        # Flag for close-loop training
+        self.cle = cle
+
+    def reset(self) -> Dict[str, np.ndarray]:
+        """
+        :param max_scene_id: the maximum scene index to sample from dataset
+        :return: the observation of first frame of sampled scene index
+        """
+
+        # Sample a episode randomly
+        self.scene_indices = [random.randint(0, self.max_scene_id)]
+        self.sim_dataset = SimulationDataset.from_dataset_indices(self.dataset, self.scene_indices, self.sim_cfg)
+
+        # Define in / outs for given scene
+        self.agents_ins_outs: DefaultDict[int, List[List[UnrollInputOutput]]] = defaultdict(list)
+        self.ego_ins_outs: DefaultDict[int, List[UnrollInputOutput]] = defaultdict(list)
+
+        # Reset CLE evaluator
+        self.reward.reset()
+
+        # Output first observation
+        self.frame_index = 0
+        ego_input = self.sim_dataset.rasterise_frame_batch(self.frame_index)
+        self.ego_input_dict = default_collate_numpy(ego_input[0])
+
+        # Only output the image attribute
+        obs = {'image': ego_input[0]["image"]}
+        return obs
+
+    def step(self, action: np.ndarray) -> GymStepOutput:
+        """
+        Analogous to sim_loop.unroll(scenes_to_unroll) in L5Kit
+
+        :param action: the action to perform on current state/frame
+        :return: the namedTuple comprising the (next observation, reward, done, info)
+                 based on the current action
+        """
+
+        frame_index = self.frame_index
+        next_frame_index = frame_index + 1
+        episode_over = next_frame_index == (len(self.sim_dataset) - self.future_num_frames)
+
+        # EGO
+        if not self.sim_cfg.use_ego_gt:
+            action = rescale_action(action)
+            action = convert_to_dict(action, self.future_num_frames)
```

Comment:
These two functions `rescale_action` and `convert_to_dict` are very particular of the environment, can you move these functions to be methods of the environment ? They can be private prefixed w/ `_`, as in `def _rescale_action(...)`.